File: http/src/main/java/com/androidx/http/net/HttpRequest.java
package com.androidx.http.net;

import android.os.Handler;
import android.os.Looper;

import com.androidx.http.net.listener.Callback;
import com.androidx.http.net.listener.HttpRequestListener;
import com.androidx.http.net.module.MsgModule;

import org.jetbrains.annotations.NotNull;
import org.json.JSONObject;

import java.io.IOException;
import java.net.SocketTimeoutException;
import java.util.Map;
import java.util.Objects;

import okhttp3.Call;
import okhttp3.Response;

public final class HttpRequest implements HttpRequestListener {

    private int currentConnect;
    private static final IHttpNetwork httpNetwork = HttpNetwork.builder();
    private static final Handler handler = new Handler(Looper.getMainLooper(), message -> {
        MsgModule msg = (MsgModule) message.obj;
        switch (message.what) {
            case -1:
                msg.getCallback().onFailure(msg.getMsg());
                return false;
            case 0:
                msg.getCallback().onSuccess(msg.getMsg());
                return false;
            case 1:
                msg.getCallback().onSuccess(msg.getMsg1());
                return false;
            default:
                return false;
        }
    });

    @Override
    public void getRequest(String url, Map<String, Object> map, int maxAnewCount, Callback callBack) {
        httpNetwork.getClient().newCall(httpNetwork.getRequest(url, map)).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                handler.sendMessage(handler.obtainMessage(-1, new MsgModule(e.getMessage() == null ? "" : e.getMessage(), callBack)));
                // 如果超时并未超过指定次数，则重新连接
                if (e instanceof SocketTimeoutException && currentConnect < maxAnewCount) {
                    currentConnect++;
                    httpNetwork.getClient().newCall(call.request()).enqueue(this);
                }
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                handler.sendMessage(handler.obtainMessage(0, new MsgModule(Objects.requireNonNull(response.body()).string(), callBack)));
            }
        });
    }

    @Override
    public void postRequestProto(String url, byte[] bytes, int maxAnewCount, Callback callBack) {
        httpNetwork.getClient().newCall(httpNetwork.postRequestProto(url, bytes)).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                handler.sendMessage(handler.obtainMessage(-1, new MsgModule(e.getMessage() == null ? "" : e.getMessage(), callBack)));
                // 如果超时并未超过指定次数，则重新连接
                if (e instanceof SocketTimeoutException && currentConnect < maxAnewCount) {
                    currentConnect++;
                    httpNetwork.getClient().newCall(call.request()).enqueue(this);
                }
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                handler.sendMessage(handler.obtainMessage(1, new MsgModule(Objects.requireNonNull(response.body()).bytes(), callBack)));
            }
        });
    }

    @Override
    public void postRequest(String url, Map<String, Object> map, int maxAnewCount, Callback callBack) {
        httpNetwork.getClient().newCall(httpNetwork.postRequest(url, map)).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                handler.sendMessage(handler.obtainMessage(-1, new MsgModule(e.getMessage() == null ? "" : e.getMessage(), callBack)));
                // 如果超时并未超过指定次数，则重新连接
                if (e instanceof SocketTimeoutException && currentConnect < maxAnewCount) {
                    currentConnect++;
                    httpNetwork.getClient().newCall(call.request()).enqueue(this);
                }
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                handler.sendMessage(handler.obtainMessage(0, new MsgModule(Objects.requireNonNull(response.body()).string(), callBack)));
            }
        });
    }

    @Override
    public void postRequest(String url, JSONObject json, int maxAnewCount, Callback callBack) {
        httpNetwork.getClient().newCall(httpNetwork.postRequest(url, json)).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                handler.sendMessage(handler.obtainMessage(-1, new MsgModule(e.getMessage() == null ? "" : e.getMessage(), callBack)));
                // 如果超时并未超过指定次数，则重新连接
                if (e instanceof SocketTimeoutException && currentConnect < maxAnewCount) {
                    currentConnect++;
                    httpNetwork.getClient().newCall(call.request()).enqueue(this);
                }
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                handler.sendMessage(handler.obtainMessage(0, new MsgModule(Objects.requireNonNull(response.body()).string(), callBack)));
            }
        });
    }

    @Override
    public void deleteRequest(String url, JSONObject json, int maxAnewCount, Callback callBack) {
        httpNetwork.getClient().newCall(httpNetwork.deleteRequest(url, json)).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                handler.sendMessage(handler.obtainMessage(-1, new MsgModule(e.getMessage() == null ? "" : e.getMessage(), callBack)));
                // 如果超时并未超过指定次数，则重新连接
                if (e instanceof SocketTimeoutException && currentConnect < maxAnewCount) {
                    currentConnect++;
                    httpNetwork.getClient().newCall(call.request()).enqueue(this);
                }
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                handler.sendMessage(handler.obtainMessage(0, new MsgModule(Objects.requireNonNull(response.body()).string(), callBack)));
            }
        });
    }

    @Override
    public void deleteRequest(String url, Map<String, Object> map, int maxAnewCount, Callback callBack) {
        httpNetwork.getClient().newCall(httpNetwork.deleteRequest(url, map)).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                handler.sendMessage(handler.obtainMessage(-1, new MsgModule(e.getMessage() == null ? "" : e.getMessage(), callBack)));
                // 如果超时并未超过指定次数，则重新连接
                if (e instanceof SocketTimeoutException && currentConnect < maxAnewCount) {
                    currentConnect++;
                    httpNetwork.getClient().newCall(call.request()).enqueue(this);
                }
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                handler.sendMessage(handler.obtainMessage(0, new MsgModule(Objects.requireNonNull(response.body()).string(), callBack)));
            }
        });
    }

    @Override
    public void formRequest(String url, JSONObject json, int maxAnewCount, Callback callBack) {
        httpNetwork.getClient().newCall(httpNetwork.formRequest(url, json)).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                handler.sendMessage(handler.obtainMessage(-1, new MsgModule(e.getMessage() == null ? "" : e.getMessage(), callBack)));
                // 如果超时并未超过指定次数，则重新连接
                if (e instanceof SocketTimeoutException && currentConnect < maxAnewCount) {
                    currentConnect++;
                    httpNetwork.getClient().newCall(call.request()).enqueue(this);
                }
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                handler.sendMessage(handler.obtainMessage(0, new MsgModule(Objects.requireNonNull(response.body()).string(), callBack)));
            }
        });
    }

}

File: http/src/main/java/com/androidx/http/net/HttpRequest.kt
package com.androidx.http.net

import android.os.Handler
import android.os.Looper
import android.os.Message
import com.androidx.http.net.listener.BytesCallback
import com.androidx.http.net.listener.HttpRequestListener
import com.androidx.http.net.listener.StringCallback
import com.androidx.http.net.module.MsgModule
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Response
import org.json.JSONObject
import java.io.IOException
import java.net.SocketTimeoutException
import java.util.*

class HttpRequest : HttpRequestListener {

    private var currentConnect = 0
    private val httpNetwork: IHttpNetwork = HttpNetwork.builder()
    private val handler = Handler(Looper.getMainLooper(), Handler.Callback label@{ message: Message ->
            val msg = message.obj as MsgModule
            when (message.what) {
                -2 -> {
                    msg.bytesCallback.onFailure(String(msg.msg1))
                    return@label false
                }
                -1 -> {
                    msg.stringCallback.onFailure(msg.msg)
                    return@label false
                }
                0 -> {
                    msg.stringCallback.onSuccess(msg.msg)
                    return@label false
                }
                1 -> {
                    msg.bytesCallback.onSuccess(msg.msg1)
                    return@label false
                }
                else -> return@label false
            }
        })

    override fun getRequest(
        url: String?,
        map: MutableMap<String, Any>?,
        maxAnewCount: Int,
        callBack: StringCallback?
    ) {
        httpNetwork.client.newCall(httpNetwork.getRequest(url, map)).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                handler.sendMessage(
                    handler.obtainMessage(
                        -1, MsgModule(
                            (if (e.message == null) "" else e.message)!!,
                            callBack!!
                        )
                    )
                )
                // 如果超时并未超过指定次数，则重新连接
                if (e is SocketTimeoutException && currentConnect < maxAnewCount) {
                    currentConnect++
                    httpNetwork.client.newCall(call.request()).enqueue(this)
                }
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                handler.sendMessage(
                    handler.obtainMessage(
                        0, MsgModule(
                            Objects.requireNonNull(response.body)!!.string(),
                            callBack!!
                        )
                    )
                )
            }
        })
    }

    override fun postRequestProto(
        url: String?,
        bytes: ByteArray?,
        maxAnewCount: Int,
        callBack: BytesCallback?
    ) {
        httpNetwork.client.newCall(httpNetwork.postRequestProto(url, bytes))
            .enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    handler.sendMessage(
                        handler.obtainMessage(
                            -2, MsgModule(
                                e.message!!.toByteArray(),
                                callBack!!
                            )
                        )
                    )
                    // 如果超时并未超过指定次数，则重新连接
                    if (e is SocketTimeoutException && currentConnect < maxAnewCount) {
                        currentConnect++
                        httpNetwork.client.newCall(call.request()).enqueue(this)
                    }
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    handler.sendMessage(
                        handler.obtainMessage(
                            1, MsgModule(
                                Objects.requireNonNull(response.body)!!.bytes(), callBack!!
                            )
                        )
                    )
                }
            })
    }

    override fun postRequest(
        url: String?,
        map: MutableMap<String, Any>?,
        maxAnewCount: Int,
        callBack: StringCallback?
    ) {
        httpNetwork.client.newCall(httpNetwork.postRequest(url, map))
            .enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    handler.sendMessage(
                        handler.obtainMessage(
                            -1, MsgModule(
                                (if (e.message == null) "" else e.message)!!,
                                callBack!!
                            )
                        )
                    )
                    // 如果超时并未超过指定次数，则重新连接
                    if (e is SocketTimeoutException && currentConnect < maxAnewCount) {
                        currentConnect++
                        httpNetwork.client.newCall(call.request()).enqueue(this)
                    }
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    handler.sendMessage(
                        handler.obtainMessage(
                            0, MsgModule(
                                Objects.requireNonNull(response.body)!!.string(), callBack!!
                            )
                        )
                    )
                }
            })
    }

    override fun postRequest(
        url: String?,
        json: JSONObject?,
        maxAnewCount: Int,
        callBack: StringCallback?
    ) {
        httpNetwork.client.newCall(httpNetwork.postRequest(url, json))
            .enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    handler.sendMessage(
                        handler.obtainMessage(
                            -1, MsgModule(
                                (if (e.message == null) "" else e.message)!!,
                                callBack!!
                            )
                        )
                    )
                    // 如果超时并未超过指定次数，则重新连接
                    if (e is SocketTimeoutException && currentConnect < maxAnewCount) {
                        currentConnect++
                        httpNetwork.client.newCall(call.request()).enqueue(this)
                    }
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    handler.sendMessage(
                        handler.obtainMessage(
                            0, MsgModule(
                                Objects.requireNonNull(response.body)!!.string(), callBack!!
                            )
                        )
                    )
                }
            })
    }

    override fun deleteRequest(
        url: String?,
        json: JSONObject?,
        maxAnewCount: Int,
        callBack: StringCallback?
    ) {
        httpNetwork.client.newCall(httpNetwork.deleteRequest(url, json))
            .enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    handler.sendMessage(
                        handler.obtainMessage(
                            -1, MsgModule(
                                (if (e.message == null) "" else e.message)!!,
                                callBack!!
                            )
                        )
                    )
                    // 如果超时并未超过指定次数，则重新连接
                    if (e is SocketTimeoutException && currentConnect < maxAnewCount) {
                        currentConnect++
                        httpNetwork.client.newCall(call.request()).enqueue(this)
                    }
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    handler.sendMessage(
                        handler.obtainMessage(
                            0, MsgModule(
                                Objects.requireNonNull(response.body)!!.string(), callBack!!
                            )
                        )
                    )
                }
            })
    }

    override fun deleteRequest(
        url: String?,
        map: MutableMap<String, Any>?,
        maxAnewCount: Int,
        callBack: StringCallback?
    ) {
        httpNetwork.client.newCall(httpNetwork.deleteRequest(url, map))
            .enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    handler.sendMessage(
                        handler.obtainMessage(
                            -1, MsgModule(
                                (if (e.message == null) "" else e.message)!!,
                                callBack!!
                            )
                        )
                    )
                    // 如果超时并未超过指定次数，则重新连接
                    if (e is SocketTimeoutException && currentConnect < maxAnewCount) {
                        currentConnect++
                        httpNetwork.client.newCall(call.request()).enqueue(this)
                    }
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    handler.sendMessage(
                        handler.obtainMessage(
                            0, MsgModule(
                                Objects.requireNonNull(response.body)!!.string(), callBack!!
                            )
                        )
                    )
                }
            })
    }

    override fun formRequest(
        url: String?,
        json: JSONObject?,
        maxAnewCount: Int,
        callBack: StringCallback?
    ) {
        httpNetwork.client.newCall(httpNetwork.formRequest(url, json))
            .enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    handler.sendMessage(
                        handler.obtainMessage(
                            -1, MsgModule(
                                (if (e.message == null) "" else e.message)!!,
                                callBack!!
                            )
                        )
                    )
                    // 如果超时并未超过指定次数，则重新连接
                    if (e is SocketTimeoutException && currentConnect < maxAnewCount) {
                        currentConnect++
                        httpNetwork.client.newCall(call.request()).enqueue(this)
                    }
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    handler.sendMessage(
                        handler.obtainMessage(
                            0, MsgModule(
                                Objects.requireNonNull(response.body)!!.string(), callBack!!
                            )
                        )
                    )
                }
            })
    }
}